flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/RemoveUnreachableCoalesceArgumentsRule.java

```diff
@@ -0,0 +1,188 @@
+ */
+@Internal
+public class RemoveUnreachableCoalesceArgumentsRule
+        extends RelRule<RemoveUnreachableCoalesceArgumentsRule.Config> {
+
+    public static final RelOptRule PROJECT_INSTANCE =
+            Config.EMPTY.as(Config.class).withProject().toRule();
+    public static final RelOptRule FILTER_INSTANCE =
+            Config.EMPTY.as(Config.class).withFilter().toRule();
+    public static final RelOptRule JOIN_INSTANCE =
+            Config.EMPTY.as(Config.class).withJoin().toRule();
+    public static final RelOptRule CALC_INSTANCE =
+            Config.EMPTY.as(Config.class).withCalc().toRule();
+
+    private static final UnreachableCoalesceArgumentsRemoveRexShuttle REX_SHUTTLE_INSTANCE =
+            new UnreachableCoalesceArgumentsRemoveRexShuttle();
+
+    public RemoveUnreachableCoalesceArgumentsRule(Config config) {
+        super(config);
+    }
+
+    @Override
+    public void onMatch(RelOptRuleCall call) {
+        final RelNode relNode = call.rel(0);
+        call.transformTo(relNode.accept(REX_SHUTTLE_INSTANCE));
+    }
+
+    private static class UnreachableCoalesceArgumentsRemoveRexShuttle extends RexShuttle {
+
+        @Override
+        public RexNode visitCall(RexCall call) {
+            call = (RexCall) super.visitCall(call);
+
+            // Not a coalesce invocation, skip it
+            if (!call.getOperator()
+                    .getName()
+                    .equals(BuiltInFunctionDefinitions.COALESCE.getName())) {
+                return call;
+            }
+
+            int firstNonNullableArgIndex = getFirstNonNullableArgumentIndex(call);
+
+            // If it's the first argument, just return the argument without the coalesce invocation
+            if (firstNonNullableArgIndex == 0) {
+                return call.operands.get(0);
+            }
+
+            // If it's the last argument, or no non-null argument was found, return the original
+            // call
+            if (firstNonNullableArgIndex == call.operands.size() - 1
+                    || firstNonNullableArgIndex == -1) {
+                return call;
+            }
+
+            // Return the coalesce invocation with a trimmed argument list
+            final List<RexNode> trimmedOperandsList =
+                    call.operands.subList(0, firstNonNullableArgIndex + 1);
+            return call.clone(call.getType(), trimmedOperandsList);
+        }
+
+        private int getFirstNonNullableArgumentIndex(RexCall call) {
+            for (int argIndex = 0; argIndex < call.operands.size(); argIndex++) {
+                if (!call.operands.get(argIndex).getType().isNullable()) {
+                    return argIndex;
+                }
+            }
+            return -1;
+        }
+    }
+
+    private static boolean hasCoalesceInvocation(RexNode node) {
+        return FlinkRexUtil.hasOperatorCallMatching(
+                node, op -> op.getName().equals(BuiltInFunctionDefinitions.COALESCE.getName()));
```

Comment:
It'd actually be nice not to have to match on the name. I think the operator here will be a `BridgingSqlFunction`, and then we could actually look into it and compare on the actual function definition (those implement `equals`).